Mesh generators for 1D B Spline

lipohlmann opened the pull request:
closes #31003 

## Reason
Part of larger scope of work looking to create a pipe connector meshing utility to connect two meshes.

## Design
Uses two points with two directions to draw a B-spline. Additional sharpness parameter created to allow for additional flexibility in the curve's shape. Generator is written such that the user does not need to explicitly specify control points -- the creation of these is handled under the hood based on some user input. Spline is designed to honor the directions specified to create the smoothest curve possible.

## Impact
Creates a new 1D meshed curve. Will eventually be used for mesh extrusion.

Commit:
Add mesh generator for creating a BSpline curve edge mesh

first draft of curve generator

Fixups on BSpline generator class
refs #31003

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3031531/), step Clang format on 61e670f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31004/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31004/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c0da5d689621e363383a7d0a7776f7d7af217f5a`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3714716/), step Docs: sync website on 77463de wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31004/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3714754/), step Generate coverage on 77463de wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>23ccd6</th>
      <th colspan=4>#31004 77463d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.87%</td>
      <td>85.88%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>130828</td>
      <td>130900</td>
      <td>+72</td>
      <td>76</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>21526</td>
      <td>21530</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31004/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31004/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/reactor)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31004/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3644012/), step Results summary on 5c3023e wanted to post the following:

## Framework test summary

Compared against aec0a32 in job [civet.inl.gov/job/3640883](https://civet.inl.gov/job/3640883).

### Added tests

| Test                                                                     | Time (s)   | Memory (MB)   |
|--------------------------------------------------------------------------|------------|---------------|
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpnessError1` | 0.69       | 0.00          |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree4`         | 0.63       | 0.00          |
| `meshgenerators/bspline_curve_generator.3D_curve_arbitrary`              | 0.62       | 112.99        |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_cpsError`        | 0.61       | 0.00          |
| `meshgenerators/bspline_curve_generator.circle_error1`                   | 0.61       | 72.77         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpnessError2` | 0.61       | 108.04        |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree3`         | 0.59       | 0.14          |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness25`     | 0.59       | 48.44         |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree2`         | 0.58       | 0.15          |
| `meshgenerators/bspline_curve_generator.circle_error2`                   | 0.58       | 111.20        |
| `meshgenerators/bspline_curve_generator.3D_circle`                       | 0.58       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_circle`                       | 0.58       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_curve_s_curve`                | 0.56       | 88.91         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_edge3`           | 0.56       | 22.62         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_edge4`           | 0.55       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness100`    | 0.54       | 75.92         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple`                 | 0.54       | 105.41        |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness50`     | 0.53       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness75`     | 0.52       | 74.14         |
| `meshgenerators/bspline_curve_generator.3D_curve_simple`                 | 0.43       | 0.00          |

### Run time changes

| Test                                                            | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|-----------------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `convergence/reference_residual_convergence.zero_tolerance_ref` | 4.28       | 6.92       | +61.63% | 124.63      | 121.15      |
| `problems/reference_residual_problem.zero_tolerance_ref`        | 4.28       | 6.86       | +60.45% | 111.74      | 113.37      |
| `executioners/multisys-ss-detection.multi-sys-ss`               | 4.99       | 7.49       | +50.26% | 136.29      | 123.24      |
## Modules test summary

Compared against aec0a32 in job [civet.inl.gov/job/3640883](https://civet.inl.gov/job/3640883).

### No added tests


### Run time changes

| Test                                                                  | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|-----------------------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `stochastic_tools/test:web_server_control.stochastic_control/normal`  | 2.71       | 4.35       | +60.82% | 264.14      | 269.27      |
| `solid_mechanics/test:smeared_cracking.cracking_rotation_pres_dir_z`  | 4.14       | 6.30       | +52.29% | 135.85      | 125.10      |
| `solid_mechanics/test:smeared_cracking.cracking_rotation_pres_dir_xz` | 4.13       | 6.26       | +51.54% | 126.68      | 130.38      |
| `solid_mechanics/test:smeared_cracking.rz_exponential`                | 12.39      | 18.71      | +51.02% | 134.13      | 130.05      |
| `solid_mechanics/test:combined_creep_plasticity.creepWithPlasticity`  | 5.64       | 8.51       | +50.83% | 119.90      | 119.25      |

GiudGiud committed:
Add tests for BSpline curve generator
refs #31003

GiudGiud committed:
Improve defaults for BSpline generator
refs #31003

GiudGiud committed:
Address Guillaume s review
- cleaner switch syntax than 3x new elem alloc
- add capability for user-set subdomain ID and name
- more docstrings
- organize parameters, add groups
refs #31003

GiudGiud committed:
Add documentation for the BSpline curve generator
refs #31003

GiudGiud reviewed (approved):
@roystgnr this is PR number 3.
It has fixes for the utils that were not in PR 1 and 2 but I moved those commits now.

I addressed my review, and fixed the commit history, separating the various changes hence some commits are attributed to me

GiudGiud committed:
Unity build fix
Unit test throw fixc
refs #31003

GiudGiud committed:
Fix the test requirements:
- no duplicates
- clearer
- don't check input param logic
Avoid unnecessary num_cp warning
refs #31003

moosebuild commented:
Job [LibTorch](https://civet.inl.gov/job/3648239/) on cf098cc : invalidated by @loganharbour

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3657438/), step Results summary on a9966ef wanted to post the following:

## Framework test summary

Compared against 581bd7b in job [civet.inl.gov/job/3655665](https://civet.inl.gov/job/3655665).

### Added tests

| Test                                                                  | Time (s)   | Memory (MB)   |
|-----------------------------------------------------------------------|------------|---------------|
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree2`      | 0.64       | 0.00          |
| `meshgenerators/bspline_curve_generator.3D_curve_arbitrary`           | 0.60       | 0.00          |
| `meshgenerators/bspline_curve_generator.circle_error2`                | 0.57       | 0.00          |
| `meshgenerators/bspline_curve_generator.circle_error1`                | 0.56       | 0.00          |
| `meshgenerators/bspline_curve_generator.3D_circle`                    | 0.55       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_edge4`        | 0.54       | 87.09         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_edge3`        | 0.54       | 65.92         |
| `meshgenerators/bspline_curve_generator.3D_curve_simple`              | 0.53       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_circle`                    | 0.53       | 114.56        |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness75`  | 0.53       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_curve_simple`              | 0.53       | 111.41        |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness50`  | 0.53       | 0.00          |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree3`      | 0.53       | 108.69        |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree4`      | 0.52       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness100` | 0.52       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_cpsError`     | 0.52       | 0.00          |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness25`  | 0.51       | 0.00          |
## Modules test summary

Compared against 581bd7b in job [civet.inl.gov/job/3655665](https://civet.inl.gov/job/3655665).

### No added tests


### Run time changes

| Test                                                                      | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|---------------------------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `stochastic_tools/test:web_server_control.stochastic_control/batch_reset` | 2.64       | 4.02       | +52.25% | 240.82      | 245.22      |

GiudGiud committed:
Address code review

- add more details
- error on unsupported parallel line case (for which sharpness isnt defined)
- document end direction
refs #31003
Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

GiudGiud commented on framework/doc/content/source/meshgenerators/BSplineCurveGenerator.md:
```diff
@@ -0,0 +1,16 @@
+avoiding a discontinuity in the slope.
+
+The spline shape parameters should be adapted to obtain the desired curve. The sharpness is notably defined here
+as a measure of the proximity of the spline to a curve consisting of three orthogonal segments, two continuing from
```
The directions are aligned but points are not and are not along a circle case is not supported either right now but I think we could do it fairly easily by specifying the "closest point" between the two parallel lines to be right in between the two points. 

I'm not necessarily trying to add more to this PR though

GiudGiud replied:
> Thanks, I'm just confused by the statement that the line segments are always orthogonal, 

in that construction, the three lines are orthogonal 2 by 2 in a way. The shortest path from one line to another is to build a 3rd line, orthogonal to both the other lines

joshuahansel replied:
Ah ok, the language made it sound as if they were all orthogonal to one another, but I see it's just that the shortest-path line is orthogonal to both the others.

roystgnr replied:
I'd change "three orthogonal segments" to "three line segments, each orthogonal to its neighbor(s)".

roystgnr replied:
I'd like to know, though: what happens if we have high sharpness, our end directions (considered as rays starting from their endpoints a distance D apart) are coplanar, and they're, say, 1 degree different in angle.  Those rays intersect at a distance of roughly D/tan(1°), so I'd think that sharpness=1 means we'd get two super-long line segments out to that intersection point.  (plus one length-0 degenerate middle segment, whatever).

But then if we decrease that angle from 1° to 0.0000001° or so, eventually getting below `TOLERANCE` in our comparisons, we'll suddenly snap from a curve that goes D/ε away to a semicircle that goes D/2 away?

I don't actually see anything better we could do there, but that's somewhat unexpected and the "just make a semicircle" fallback should be documented.

GiudGiud replied:
yeah these end cases are hard. I think users will be thankful for that circle though, better than just erroring out and letting them figure out an alternative generator

joshuahansel commented on framework/src/meshgenerators/BSplineCurveGenerator.C:
```diff
@@ -0,0 +1,158 @@
+  // Spline shape parameters
+  params.addParam<unsigned int>("degree", 3, "Degree of interpolating polynomial.");
+  params.addRangeCheckedParam<libMesh::Real>(
+      "sharpness", 0.6, "sharpness>0 & sharpness<=1", "Sharpness of curve bend.");
```
The doc should show how this sharpness value is used (probably should show up in an equation).

GiudGiud replied:
added in last but one commit

GiudGiud committed:
Add doco on how the sharpness comes in

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3713935/), step Results summary on d5f5d22 wanted to post the following:

## Framework test summary

Compared against 23ccd6e in job [civet.inl.gov/job/3713096](https://civet.inl.gov/job/3713096).

### Added tests

| Test                                                                  | Time (s)   | Memory (MB)   |
|-----------------------------------------------------------------------|------------|---------------|
| `meshgenerators/bspline_curve_generator.2D_curve_simple_cpsError`     | 0.57       | 71.70         |
| `meshgenerators/bspline_curve_generator.2D_circle`                    | 0.57       | 101.55        |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_edge4`        | 0.57       | 61.23         |
| `meshgenerators/bspline_curve_generator.circle_error1`                | 0.55       | 79.38         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness100` | 0.55       | 54.20         |
| `meshgenerators/bspline_curve_generator.3D_curve_simple`              | 0.54       | 63.81         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_edge3`        | 0.54       | 61.98         |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree2`      | 0.53       | 59.62         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness50`  | 0.53       | 93.41         |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree3`      | 0.53       | 159.12        |
| `meshgenerators/bspline_curve_generator.3D_curve_arbitrary`           | 0.53       | 62.76         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness75`  | 0.53       | 49.57         |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree4`      | 0.52       | 140.73        |
| `meshgenerators/bspline_curve_generator.2D_curve_simple`              | 0.52       | 31.31         |
| `meshgenerators/bspline_curve_generator.3D_circle`                    | 0.51       | 116.70        |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness25`  | 0.49       | 139.33        |
| `meshgenerators/bspline_curve_generator.circle_error2`                | 0.48       | 56.59         |
## Modules test summary

Compared against 23ccd6e in job [civet.inl.gov/job/3713096](https://civet.inl.gov/job/3713096).

No change

roystgnr reviewed (changes requested):
We ought to be documenting the semicircle fallback case, and maybe changing "three orthogonal segments" if Josh likes my wording or some other wording better, but I'm not seeing anything else I don't like here.

joshuahansel reviewed (changes requested):
Please document how sharpness value is used.

GiudGiud committed:
Address Roy's review:
- orthgonal consecutive segments
- tell user about the circle fallback in documentation not just warning

GiudGiud commented:
documentation should be updated now.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3714781/), step Results summary on 77463de wanted to post the following:

## Framework test summary

Compared against 23ccd6e in job [civet.inl.gov/job/3713096](https://civet.inl.gov/job/3713096).

### Added tests

| Test                                                                  | Time (s)   | Memory (MB)   |
|-----------------------------------------------------------------------|------------|---------------|
| `meshgenerators/bspline_curve_generator.circle_error1`                | 0.76       | 94.43         |
| `meshgenerators/bspline_curve_generator.3D_circle`                    | 0.58       | 92.81         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_edge4`        | 0.55       | 77.38         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_cpsError`     | 0.54       | 69.21         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness100` | 0.53       | 58.01         |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree4`      | 0.52       | 75.36         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness25`  | 0.52       | 65.25         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness50`  | 0.52       | 70.07         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_edge3`        | 0.52       | 100.28        |
| `meshgenerators/bspline_curve_generator.2D_curve_simple`              | 0.51       | 110.35        |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree3`      | 0.51       | 82.76         |
| `meshgenerators/bspline_curve_generator.2D_curve_simple_sharpness75`  | 0.51       | 100.03        |
| `meshgenerators/bspline_curve_generator.3D_curve_arbitrary`           | 0.50       | 57.81         |
| `meshgenerators/bspline_curve_generator.3D_curve_simple_degree2`      | 0.50       | 147.87        |
| `meshgenerators/bspline_curve_generator.3D_curve_simple`              | 0.50       | 37.29         |
| `meshgenerators/bspline_curve_generator.2D_circle`                    | 0.49       | 69.57         |
| `meshgenerators/bspline_curve_generator.circle_error2`                | 0.30       | 68.33         |
## Modules test summary

Compared against 23ccd6e in job [civet.inl.gov/job/3713096](https://civet.inl.gov/job/3713096).

No change

GiudGiud commented:
Thanks for the reviews!
